internal/repository/memory.go

```diff
@@ -0,0 +1,330 @@
+var _ repository.ListableRepository = (*InMemoryRepository)(nil)
+
+// nodeReaderWriter allows reading or writing to elements in a InMemoryRepository
+type nodeReaderWriter struct {
+	path        string
+	repo        *InMemoryRepository
+	writing     bool
+	readCursor  int
+	writeCursor int
+}
+
+// InMemoryRepository implements an in-memory version of the
+// repository.Repository type. It is useful for writing test cases, and may
+// also be of use as a template for people wanting to implement their own
+// "virtual repository". In future, we may consider moving this into the public
+// API.
+//
+// Because of it's design, this repository has several quirks:
+//
+// * The Parent() of a path that exists does not necessarily exist
+//
+// * Listing takes O(number of extant paths in the repository), rather than
+//   O(number of children of path being listed).
+//
+// This repository is not designed to be particularly fast or robust, but
+// rather to be simple and easy to read. If you need performance, look
+// elsewhere.
+//
+// Since 2.0.0
+type InMemoryRepository struct {
+	// Data is exposed to allow tests to directly insert their own data
+	// without having to go through the API
+	Data map[string][]byte
+
+	scheme string
+}
+
+// Read implements io.Reader.Read
+func (n *nodeReaderWriter) Read(p []byte) (int, error) {
+
+	// first make sure the requested path actually exists
+	data, ok := n.repo.Data[n.path]
+	if !ok {
+		return 0, fmt.Errorf("path '%s' not present in InMemoryRepository", n.path)
+	}
+
+	// copy it into p - we maintain counts since len(data) may be smaller
+	// than len(p)
+	count := 0
+	j := 0 // index into p
+	for ; (j < len(p)) && (n.readCursor < len(data)); n.readCursor++ {
+		p[j] = data[n.readCursor]
+		count++
+		j++
+	}
+
+	// generate EOF if needed
+	var err error = nil
+	if n.readCursor >= len(data) {
+		err = io.EOF
+	}
+
+	return count, err
+}
+
+// Close implements io.Closer.Close
+func (n *nodeReaderWriter) Close() error {
+	n.readCursor = 0
+	n.writeCursor = 0
+	n.writing = false
+	return nil
+}
+
+// Write implements io.Writer.Write
+//
+// This implementation automatically creates the path n.path if it does not
+// exist. If it does exist, it is overwritten.
+func (n *nodeReaderWriter) Write(p []byte) (int, error) {
+
+	// guarantee that the path exists
+	_, ok := n.repo.Data[n.path]
+	if !ok {
+		n.repo.Data[n.path] = []byte{}
+	}
+
+	// overwrite the file if we haven't already started writing to it
+	if !n.writing {
+		n.repo.Data[n.path] = make([]byte, 0)
+		n.writing = true
+	}
+
+	// copy the data into the node buffer
+	count := 0
+	start := n.writeCursor
+	for ; n.writeCursor < start+len(p); n.writeCursor++ {
+		// extend the file if needed
+		if len(n.repo.Data) < n.writeCursor+len(p) {
+			n.repo.Data[n.path] = append(n.repo.Data[n.path], 0)
+		}
+		n.repo.Data[n.path][n.writeCursor] = p[n.writeCursor-start]
+		count++
+	}
+
+	return count, nil
+}
+
+// Name implements fyne.URI*Closer.URI
```

Comment:
Asterisk?

Reply:
Ah Reader/Writer

Reply:
35ed6d4ee25ea505145b287295e9785e00aca7e4

Reply:
Yeah, I thought it was obvious, but if it made you pause, obviously not. I re-worded it to be explicit so hopefully nobody will be confused in future.